asserts/ifacedecls.go

```diff
@@ -0,0 +1,477 @@
+	case "deny-auto-connection":
+		r.DenyAutoConnection = cstrs
+	default:
+		panic("unknown PlugRule field " + field)
+	}
+}
+
+// PlugInstallationConstraints specifies a set of constraints on an interface plug relevant to the installation of snap.
+type PlugInstallationConstraints struct {
+	PlugAttributes *AttributeConstraints
+}
+
+func compilePlugInstallationConstraints(interfaceName string, entry string, constraints interface{}) (*PlugInstallationConstraints, error) {
+	context := fmt.Sprintf("%s in plug rule for interface %q", entry, interfaceName)
+	cMap, shortcut, err := checkMapOrShortcut(context, constraints)
+	if err != nil {
+		return nil, err
+	}
+	if cMap == nil {
+		if shortcut {
+			return &PlugInstallationConstraints{PlugAttributes: AlwaysMatchAttributes}, nil
+		}
+		return &PlugInstallationConstraints{PlugAttributes: NeverMatchAttributes}, nil
+	}
+	attrs, err := compileAttributeContraints(cMap["plug-attributes"])
+	if err != nil {
+		return nil, fmt.Errorf("cannot compile plug-attributes in %s: %v", context, err)
+	}
+	return &PlugInstallationConstraints{PlugAttributes: attrs}, nil
+}
+
+// PlugConnectionConstraints specfies a set of constraints on an interface plug for a snap relevant to its connection or auto-connection.
+type PlugConnectionConstraints struct {
+	SlotSnapIDs      []string
+	SlotPublisherIDs []string
+
+	PlugAttributes *AttributeConstraints
+	SlotAttributes *AttributeConstraints
+}
+
+func (c *PlugConnectionConstraints) setAttributeConstraints(field string, cstrs *AttributeConstraints) {
+	switch field {
+	case "plug-attributes":
+		c.PlugAttributes = cstrs
+	case "slot-attributes":
+		c.SlotAttributes = cstrs
+	default:
+		panic("unknown PlugConnectionConstraints field " + field)
+	}
+}
+
+func (c *PlugConnectionConstraints) setIDConstraints(field string, cstrs []string) {
+	switch field {
+	case "slot-snap-id":
+		c.SlotSnapIDs = cstrs
+	case "slot-publisher-id":
+		c.SlotPublisherIDs = cstrs
+	default:
+		panic("unknown PlugConnectionConstraints field " + field)
+	}
+}
+
+var (
+	attributeConstraints = []string{"plug-attributes", "slot-attributes"}
+	plugIDConstraints    = []string{"slot-publisher-id", "slot-snap-id"}
+)
+
+func compilePlugConnectionConstraints(interfaceName string, entry string, constraints interface{}) (*PlugConnectionConstraints, error) {
+	context := fmt.Sprintf("%s in plug rule for interface %q", entry, interfaceName)
+	cMap, shortcut, err := checkMapOrShortcut(context, constraints)
+	if err != nil {
+		return nil, err
+	}
+	if cMap == nil {
+		if shortcut {
+			return &PlugConnectionConstraints{PlugAttributes: AlwaysMatchAttributes, SlotAttributes: AlwaysMatchAttributes}, nil
+		}
+		return &PlugConnectionConstraints{PlugAttributes: NeverMatchAttributes, SlotAttributes: NeverMatchAttributes}, nil
+	}
+	plugConnCstrs := &PlugConnectionConstraints{}
+	defaultUsed := 0
+	for _, field := range plugIDConstraints {
+		// TODO: check these further against regexps?
```

Comment:
Yeah, might be a good idea to ensure we can use special keys in the future ($same or the like) without being afraid we're blowing something up.